coderd/workspacebuilds.go

```diff
@@ -673,12 +660,84 @@ func (api *API) postWorkspaceBuilds(rw http.ResponseWriter, r *http.Request) {
 			Message: "Internal error converting workspace build.",
 			Detail:  err.Error(),
 		})
-		return
+		xerrors.Errorf("Convert workspace build: %w", err.Error())
 	}
 
 	api.publishWorkspaceUpdate(ctx, workspace.ID)
 
-	httpapi.Write(ctx, rw, http.StatusCreated, apiBuild)
+	return apiBuild, nil
+}
+
+// Restarts a workspace.
+//
+// @Summary Restart workspace
+// @ID restart-workspace
+// @Security CoderSessionToken
+// @Accept json
+// @Produce json
+// @Tags Builds
+// @Param workspace path string true "Workspace ID" format(uuid)
+// @Success 200 {object} codersdk.WorkspaceBuild
+// @Router /workspaces/{workspace}/builds/restart [post]
+func (api *API) restartWorkspace(rw http.ResponseWriter, r *http.Request) {
+	ctx := r.Context()
+
+	// create a build - stop the workspace
+	build, err := api.postBuild(rw, r, codersdk.CreateWorkspaceBuildRequest{
+		Transition: codersdk.WorkspaceTransitionStop,
+	})
+	if err != nil {
+		httpapi.Write(ctx, rw, http.StatusInternalServerError, codersdk.Response{
+			Message: "Internal error stopping workspace.",
+			Detail:  err.Error(),
+		})
+		return
+	}
+
+	// create a build - start the workspace
+	build, err = api.postBuild(rw, r, codersdk.CreateWorkspaceBuildRequest{
```

Comment:
We seem to be starting the workspace before it has been fully stopped. How should I go about fixing this?

Reply:
Potentially related:  I notice in `cli/restart.go` we look for a [ProvisionerJob error](https://github.com/coder/coder/blob/main/cli/restart.go#L48); however, I'm unsure why we're doing that or how I should replicate that behavior inside this handler.